SAK-29422 Incorporate NYU's "public announcement system"

marktriggs opened the pull request:
Details here:

https://jira.sakaiproject.org/browse/SAK-29422

ottenhoff commented on pasystem/pasystem-impl/impl/src/resources/db/migration/mysql/V2__create_pasystem.sql:
```diff
@@ -0,0 +1,61 @@
+
+
+-- Banners
+CREATE TABLE pasystem_banner_alert
```
Table case needs to be consistent. If you create as lowercase, code should refer to table in lowercase.

Linux, and thus MySQL on Linux, is case-sensitive by default.

ottenhoff commented:
This looks good and the code looks clean. My main question is the inclusion of flyway here. Sakai currently uses SQL scripts to update from version to version. Moving to flyway for the entire project would be great. But I'm not sure flyway should be part of this one tool instead of doing it the way other tools do it....

marktriggs commented:
Thanks Sam,

We'll sort out the case with table names issues and update this PR in a moment.

Yeah, we're not married to Flyway.  We just used it because it made development easier for us, as we were migrating NYU's test Oracle instance with each release we made.  Happy to back it out and just use vanilla .sql if that's preferable?

marktriggs commented:
And actually, one thing I just realized the README doesn't explicitly say: if auto.ddl=true is set, our migrations will run along with everything else (so it's not a totally separate process to run them)

ottenhoff commented:
I believe in the past we have said that implementors can upgrade major Sakai versions via SQL scripts in references/docs/conversion without having auto.ddl enabled. 

I'd love to see flyway become a core part of Sakai. Maybe @ern can help us get there ....

marktriggs commented:
No sweat.  We'll fix up the case, move our .sql stuff into references/doc/conversion and update this PR.  Thanks again!

payten commented:
Hi @ottenhoff. We've fixed up the table name case issue and also pulled out Flyway. Thanks!

juanjmerono commented:
I'm wondering if it is possible to change translation files from json to properties format. Sakai is dealing with properties to manage translations in all tools, so is easy if we keep this common format in all cases. We also use transifex to manage translations and I'm afraid it is not ready to work with json format. Also using properties allows you to use ResourceLoader directly without any prevoius parsing process.

marktriggs commented:
Makes sense, thanks @juanjmerono.  I'll have a look at switching it to use ResourceLoader tomorrow.

ern commented:
Quick comment on the flyway piece, I like what you were thinking of by including it!
Though the addition of flyway would be something that is leveraged by the entire project.
I propose we discuss it at the Core Team meeting?

jonespm commented:
I think using flyaway for migrations on tools that don't use hibernate as a way of providing auto.ddl could be a good idea. Tools like kernel, assignment, announcement, dashboard, etc. Though we'd still have to provide conversions if auto.ddl is turned off. Otherwise this system seems like it could conflict against hibernate and also double the work. Having it at a central level so other tools could leverage it would make sense to me.

Also @marktriggs As far as javascript properties, I'd look at SAK-27993 for an example of how roster2 uses properties files and the jquery.i18n.properties.js to handle the i18n. This seemed like a good pattern for client side translations.

marktriggs commented:
Thanks all!  We've changed it to use ResourceLoader for I18n now and that's looking good.  We've rebased with master and fixed a few little CSS issues that arose, so should be good to go.

ottenhoff commented on pasystem/pasystem-impl/impl/src/resources/db/migration/mysql.sql:
```diff
@@ -0,0 +1,58 @@
+CREATE TABLE IF NOT EXISTS `pasystem_popup_screens` (
+  `uuid` VARCHAR(255) PRIMARY KEY,
```
I think these UUID database fields should be 36 fixed chars instead of varchar(255)

buckett commented on pasystem/pasystem-tool/tool/src/webapp/tools/sakai.pasystem.xml:
```diff
@@ -0,0 +1,16 @@
+          description="For managing banner alerts and popup screens">
+
+        <!--  This tool will be made available to the following site types via Site Info > Edit Tools -->
+        <category name="course"/>
```
Is the normal configuration that the PA system will be in course sites? I was expecting it to be in the administration workspace by default?

buckett commented on pasystem/pasystem-api/api/src/java/org/sakaiproject/pasystem/api/Acknowledger.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public interface Acknowledger {
+
+    public final String TEMPORARY = "temporary";
```
Switch to an enum?

marktriggs commented:
@ottenhoff Switched to CHAR(36) across the board now.  Thanks!

@buckett Putting the tool under the admin workspace is what we've been doing, yeah.  I've changed the tool registration to limit to that.  We gave the tool a separate permission (pasystem.manage) to leave the option open for non-admin users to be given access to manage notifications, but I guess changing the tool registration doesn't preclude that.

And I've switched the acknowledgement types to use an enum too.  Thanks for the suggestion.

buckett commented:
@marktriggs Yep, using the permission is nice. The "normal" Sakai way todo this would be to use the placement of the tool to work out what site todo the permission check in, rather than hard coding all checks to be against /site/!admin. I don't think this needs to be changed for this pull request to get merged as most places will just deploy this to the !admin worksite.

marktriggs commented:
Ah!  thanks @buckett.  I'd intended to go back and remove that hard-coded /site/!admin anyway, so I've just fixed that up.  It's still embedded in the EntityProvider we added for managing notifications via script, but I'm less worried about that one since it's more of a special case.

steveswinsburg commented on pasystem/pasystem-impl/impl/src/java/org/sakaiproject/pasystem/impl/banners/BannerStorage.java:
```diff
@@ -0,0 +1,264 @@
+                "  (dismissed.state = ?))" +
+
+                " ORDER BY start_time");
+
```
Has this SQL been tested on both MySQL and Oracle? It looks fine though.

marktriggs replied:
MySQL, Oracle and hsql :D

steveswinsburg commented on pasystem/pasystem-impl/impl/src/java/org/sakaiproject/pasystem/impl/common/DB.java:
```diff
@@ -0,0 +1,82 @@
+        } catch (SQLException e) {
+            throw new RuntimeException("Failure in database action: " + actionDescription, e);
+        }
+    }
```
]I'm normally an advocate of Spring JDBC since raw JDBC is really low level, but this looks fine to me :+1:

steveswinsburg commented on pasystem/pasystem-impl/impl/src/java/org/sakaiproject/pasystem/impl/popups/PopupForUser.java:
```diff
@@ -0,0 +1,142 @@
+
+                // Or was dismissed temporarily, but some time has passed
+                "  (dismissed.state = 'temporary' AND" +
+                "   (? - dismissed.dismiss_time) >= ?))");
```
Just a note about checking the SQL for both DB's.

steveswinsburg commented on pasystem/pasystem-impl/impl/src/resources/db/migration/oracle.sql:
```diff
@@ -0,0 +1,73 @@
+WHEN NOT MATCHED THEN
+INSERT (function_key, function_name)
+VALUES (SAKAI_REALM_FUNCTION_SEQ.NEXTVAL, t.function_name);
+
```
I love Oracle's merge. Nice. :100:

marktriggs commented:
Rebased against master to resolve merge conflicts due to recent portal changes.

ottenhoff commented:
Okay, this is looking good, let's merge!

steveswinsburg commented:
Agreed, and done!
